src/website/shared/management/commands/update_groups.py

```diff
@@ -0,0 +1,77 @@
+import argparse
+import logging
+from typing import Any
+
+from django.contrib.auth.models import Group, User
+from django.core.management.base import BaseCommand
+from django.db import transaction
+from shared.templatetags.gh_tags import get_gh_organization, get_gh_team
+
+# TODO: move to shared.auth or shared.auth_utils
+
+logger = logging.getLogger(__name__)
+
+
+def get_team_member_ids(orgname: str, teamname: str) -> set[int]:
+    org = get_gh_organization(orgname)
+    if org:
+        team = get_gh_team(org, teamname)
+        if team:
+            members = team.get_members()
+            logger.info(
+                f"Caching {members.totalCount} IDs from team {orgname}/{teamname}..."
+            )
+
+            # The iterator will make the extra page API calls for us.
+            return {member.id for member in members}
+    return set()
+
+
+def get_github_ids_cache() -> dict[str, set[int]]:
+    ids: dict[str, set[int]] = dict()
+
+    ids["security_team"] = get_team_member_ids("NixOS", "security")
+    ids["committers"] = get_team_member_ids("NixOS", "nixpkgs-committers")
+    ids["maintainers"] = get_team_member_ids("NixOS", "nixpkgs-maintainers")
+
+    logger.info("Done caching IDs from Github.")
+
+    return ids
+
+
+class Command(BaseCommand):
+    help = "Update group memberships according to Github memberships."
+
+    def add_arguments(self, parser: argparse.ArgumentParser) -> None:
+        pass
+
+    def handle(self, *args: str, **kwargs: Any) -> None:  # pyright: ignore reportUnusedVariable
+        logger.info("Resetting group permissions")
+
+        id_cache: dict[str, set[int]] = get_github_ids_cache()
+
+        # Get the group objects for the transaction
+        group_objects: dict[str, Group] = {}
+        for groupname in id_cache.keys():
+            group_objects[groupname] = Group.objects.get(name=groupname)
+
+        logger.info("Using Github ID cache to update database groups...")
+
+        users = User.objects.all()
+        for user in users:
+            social = user.socialaccount_set.filter(provider="github").first()  # type: ignore
+            if social:
+                # Open a single transaction for the db
+                with transaction.atomic():
```

Comment:
What does it serve to be atomic over each user instead of being atomic over all the update?

Reply:
Hmm, I remember thinking on the spot about the inconvenience of a partial update of permissions for any given user. However, with the caching being eventually set up, there's less of a reason to not make the whole update atomic.

I'll change it.

Reply:
Changed in [aa398c9](https://github.com/Nix-Security-WG/nix-security-tracker/pull/159/commits/aa398c923360d17d9ef2bf4f39af74f51212289a)